trees/utils/StrongTreeFairOCT.py

```diff
@@ -0,0 +1,284 @@
+                             for i in p_prime_df.index))) <= self.fairness_bound)
+
+                    self.model.addConstr(
+                        ((1 / count_p) * quicksum(quicksum(self.zeta[i, n, self.positive_class] for n in
+                                                                  (self.tree.Leaves + self.tree.Nodes))
+                                                         for i in p_df.index)) - (
+                                (1 / count_p_prime) * quicksum(
+                            quicksum(self.zeta[i, n, self.positive_class] for n in
+                                     self.tree.Leaves + self.tree.Nodes)
+                            for i in p_prime_df.index)) >= -1 * self.fairness_bound)
+
+        if self.fairness_type == "EOpp":
+
+            # Loop through all possible combinations of the protected feature
+            for protected_feature in self.P_col_labels:
+                for combo in combinations(self.X_p[protected_feature].unique(), 2):
+                    p = combo[0]
+                    p_prime = combo[1]
+
+                #
+                p_df = self.X_p[(self.X_p[protected_feature] == p )& (self.X_p['y'] == self.positive_class)]
+                p_prime_df = self.X_p[(self.X_p[protected_feature] == p_prime) & (self.X_p['y'] == self.positive_class)]
+
+                # Count how many samples correspond to each group
+                count_p = p_df.shape[0]
+                count_p_prime = p_prime_df.shape[0]
+
+                if count_p != 0 and count_p_prime != 0:
+                    # Sum(Sum(zeta(i,n,positive_class) for n in nodes) for i in datapoints) * 1 / (Count of Protected)
+                    self.model.addConstr(
+                        ((1 / count_p) * quicksum(quicksum(self.zeta[i, n, self.positive_class] for n in
+                                                                  self.tree.Leaves + self.tree.Nodes)
+                                                         for i in p_df.index) -
+                         ((1 / count_p_prime) * quicksum(
+                             quicksum(self.zeta[i, n, self.positive_class] for n in
+                                      self.tree.Leaves + self.tree.Nodes)
+                             for i in p_prime_df.index))) <= self.fairness_bound)
+
+                    self.model.addConstr(
+                        ((1 / count_p) * quicksum(quicksum(self.zeta[i, n, self.positive_class] for n in
+                                                                  (self.tree.Leaves + self.tree.Nodes))
+                                                         for i in p_df.index)) - (
+                                (1 / count_p_prime) * quicksum(
+                            quicksum(self.zeta[i, n, self.positive_class] for n in
+                                     self.tree.Leaves + self.tree.Nodes)
+                            for i in p_prime_df.index)) >= -1 * self.fairness_bound)
```

Comment:
It appears there is quite a bit of duplicated code here. From a quick glance it looks like the code for adding the constraints is the same for all fairness types and the key difference is how we define `p_df` and `p_prime_df`. Could we do the if statements at that part? I.e., when we are looping over combinations of `p` and `p_prime`, that is where we have the if statements